org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/folding/DefaultFoldingRangeProvider.java

```diff
@@ -0,0 +1,141 @@
+/*******************************************************************************
+ * Copyright (c) 2010 Michael Clay and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package org.eclipse.xtext.ide.editor.folding;
+
+import java.util.Collection;
+
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.emf.common.util.TreeIterator;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.TerminalRule;
+import org.eclipse.xtext.documentation.impl.AbstractMultiLineCommentProvider;
+import org.eclipse.xtext.nodemodel.ICompositeNode;
+import org.eclipse.xtext.nodemodel.INode;
+import org.eclipse.xtext.nodemodel.util.NodeModelUtils;
+import org.eclipse.xtext.parser.IParseResult;
+import org.eclipse.xtext.resource.ILocationInFileProvider;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.util.ITextRegion;
+
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+
+/**
+ * @author Michael Clay - Initial contribution and API
+ * @author Sebastian Zarnekow - Introduced FoldedRegion, use ILocationInFileProvider
+ * @author Mark Sujew - Ported to IDE project
+ * 
+ * @since 2.25
+ */
+public class DefaultFoldingRangeProvider implements IFoldingRangeProvider {
+
+	@Inject
+	private ILocationInFileProvider locationInFileProvider;
+
+	@Inject(optional = true)
+	@Named(AbstractMultiLineCommentProvider.RULE)
+	protected String ruleName = "ML_COMMENT";
+
+	@Override
+	public Collection<FoldingRange> getFoldingRanges(final XtextResource xtextDocument,
+			CancelIndicator cancelIndicator) {
+		Collection<FoldingRange> result = Sets.newLinkedHashSet();
+		IFoldingRangeAcceptor<?> foldingRegionAcceptor = createAcceptor(xtextDocument, result);
+		computeObjectFolding(xtextDocument, foldingRegionAcceptor, cancelIndicator);
+		computeCommentFolding(xtextDocument, foldingRegionAcceptor);
+		return result;
+	}
+
+	protected IFoldingRangeAcceptor<?> createAcceptor(XtextResource xtextDocument,
+			Collection<FoldingRange> foldedPositions) {
+		return new DefaultFoldingRangeAcceptor(foldedPositions);
```

Comment:
The returned acceptor is of type FoldingRangeAcceptor<?> which implies that the accept methods that take a `T additionalParam` cannot be called. What is their purpose?

Reply:
True, that doesn't make a lot of sense. I'll refactor some methods of the `DefaultFoldingRangeProvider` into an `AbstractFoldingRangeProvider<T>` that makes more use of the type parameter.